openedx/openedx-platform
# [BD-04] Convert LibraryContent XModule to XBlock

Converts the LibraryContent XModule into an XBlock.

Part of [XModule to XBlock Conversion work](https://openedx.atlassian.net/wiki/spaces/AC/pages/1472790755/XModule+to+XBlock+Conversion). 

**Testing instructions**:

1. In a course set up test Randomized Content component (code name is LibraryContentModule). See [how to](https://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/exercises_tools/randomized_content_blocks.html) for set up instructions.
2. Attempt the problems from LMS.
3. Make an export of the course.
4. Update devstack with this code.
   - Cherry-pick this commit.
   - In both `make lms-shell` and `make studio-shell`, install the updated version of the `xmodule` package with `pip install -e /common/lib/xmodule`.
   - Do `make lms-static` and `make studio-static`.
   - Do `make lms-restart` and `make studio-restart`.
5. Test the existing components in both Studio and LMS.
6. Try to add another one.
7. Import the exported archive into a new course and check that it still works the same.


**MRO Analysis**

Descriptor MRO | Module MRO | Block MRO | Notes
------------ | ------------- | ---------------- | ------------
<class 'xmodule.library_content_module.LibraryContentDescriptor'> | <class 'xmodule.library_content_module.LibraryContentModule'> | <class 'xmodule.library_content_module.LibraryContentBlock'> | Merged two classes together.
<class 'xmodule.library_content_module.LibraryContentFields'> | <class 'xmodule.library_content_module.LibraryContentFields'> | —  | Merged in LibraryContentBlock
<class 'xmodule.mako_module.MakoModuleDescriptor'> | —  | —  | Inherits from (MakoTemplateBlockBase, XModuleDescriptor) and has the get_html() method which has been implemented in the XBlock.
<class 'xmodule.mako_module.MakoTemplateBlockBase'> | —  | <class 'xmodule.mako_module.MakoTemplateBlockBase'> | — 
<class 'xmodule.xml_module.XmlDescriptor'> | —  | —  | Inherits from (XmlMixin, XModuleDescriptor) and doens't have any atttributes so we can remove it.
<class 'xmodule.xml_module.XmlMixin'> | —  | <class 'xmodule.xml_module.XmlMixin'> | — 
<class 'xmodule.xml_module.XmlParserMixin'> | —  | <class 'xmodule.xml_module.XmlParserMixin'> | — 
<class 'xmodule.x_module.XModuleDescriptor'> | —  | —  | Not needed.
<class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> | —  | <class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> | — 
 — | <class 'xmodule.x_module.XModule'> | —  | Not needed.
 — | <class 'xmodule.x_module.XModuleToXBlockMixin'> | <class 'xmodule.x_module.XModuleToXBlockMixin'> | — 
<class 'xmodule.x_module.HTMLSnippet'> | <class 'xmodule.x_module.HTMLSnippet'> | <class 'xmodule.x_module.HTMLSnippet'> | — 
<class 'xmodule.x_module.ResourceTemplates'> | —  | <class 'xmodule.x_module.ResourceTemplates'> | — 
<class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> | — 
<class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> | — 
<class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> | — 
<class 'xmodule.studio_editable.StudioEditableDescriptor'> | <class 'xmodule.studio_editable.StudioEditableModule'> | <class 'xmodule.studio_editable.StudioEditableBlock'> | —
